crates/holochain/src/core/workflow/countersigning_workflow/complete.rs

```diff
@@ -0,0 +1,232 @@
+use crate::conductor::space::Space;
+use crate::core::queue_consumer::TriggerSender;
+use crate::core::ribosome::weigh_placeholder;
+use crate::core::workflow::{WorkflowError, WorkflowResult};
+use holo_hash::{ActionHash, AgentPubKey, DhtOpHash, EntryHash};
+use holochain_keystore::AgentPubKeyExt;
+use holochain_p2p::HolochainP2pDnaT;
+use holochain_sqlite::db::ReadAccess;
+use holochain_sqlite::error::DatabaseResult;
+use holochain_state::integrate::authored_ops_to_dht_db_without_check;
+use holochain_state::mutations;
+use holochain_state::prelude::{current_countersigning_session, SourceChainResult, Store};
+use holochain_types::dht_op::ChainOp;
+use holochain_zome_types::prelude::SignedAction;
+use rusqlite::{named_params, Transaction};
+use std::sync::Arc;
+
+pub(crate) async fn inner_countersigning_session_complete(
+    space: Space,
+    network: Arc<impl HolochainP2pDnaT>,
+    author: AgentPubKey,
+    signed_actions: Vec<SignedAction>,
+    integration_trigger: TriggerSender,
+    publish_trigger: TriggerSender,
+) -> WorkflowResult<Option<EntryHash>> {
+    let authored_db = space.get_or_create_authored_db(author.clone())?;
+
+    // Using iterators is fine in this function as there can only be a maximum of 8 actions.
+    let (this_cells_action_hash, entry_hash) = match signed_actions
+        .iter()
+        .find(|a| *a.author() == author)
+        .and_then(|sa| {
+            sa.entry_hash()
+                .cloned()
+                .map(|eh| (ActionHash::with_data_sync(sa), eh))
+        }) {
+        Some(a) => a,
+        None => return Ok(None),
+    };
```

Comment:
I encountered this before understanding what the computed hash is for, but I'm just checking that this is correct. This finds the first action by the given author, and if that first action contains no entry data, this function returns `None`, even if there are other actions in the list which contain entry data. Just making sure that you don't want that `find` to be a `filter` instead, so you can catch the first action with entry data, even if there are other actions preceding it which don't contain entry data.

Reply:
Ah, I see this code was just moved. It would help to have a map of what code was modified and what was merely moved, so I don't re-review stuff.

Reply:
For the modules that live under the workflow:
- `accept` was split from the host function
- `complete` was the guts of the `countersigning_success` function that used to receive signature bundles
- `refresh` is new
- `incomplete` is new
- `success` is where the old `countersigning_success` was but does a different job, where it just updates state to gather signatures and lets the workflow do the old success behaviour in `complete`

Entirely valid to question code that was already there though. I think it's okay though. We require a bundle of signed actions from a witness. A signed action without an entry hash cannot be for a countersigning commit so it's right to reject the input in that case. The bundle should also contain exactly one action for each participant so it would be invalid to receive a bundle with multiple matches for the same author. We could check that here, but the length of the bundle is checked against the number of participants in the session later, so I think we're covering all our bases here for rejecting invalid input and not missing useful data.